### PR TITLE
Migrate PlusRemote from Python to C++

### PR DESCRIPTION
**Pull request by Sunderlandkyl**
@lassoan I think this should be it for converting PlusRemote to C++

Before this is merged, I would like to make sure that all affected modules are changed as well.

TODO:

- [x] SlicerIGT - Guidelet: https://github.com/SlicerIGT/SlicerIGT/pull/179
- [x] ~~PerkLab/SpineFusionNav: https://github.com/PerkLab/SpineFusionNav/blob/master/SpineMain/SpineNavLib/NavigationWidget.py~~ -- Not currently active

Requires new macros from:
- https://github.com/Slicer/Slicer/pull/973
- https://gitlab.kitware.com/vtk/vtk/merge_requests/4427

### CODE REVIEW

**Comment by Sunderlandkyl** [comment]
@lassoan I couldn't find any other uses of PlusRemote in any of the PerkLab/SlicerIGT/PerkTutor repositories. 
Grace said that the SpineFusionNav is not currently active, so that doesn't need to be changed at this time.

Would you mind reviewing this to see if there are any issues?

File: PlusRemote/Logic/vtkSlicerPlusRemoteLogic.cxx [hunk 1]
```diff
@@ -0,0 +1,580 @@
...
+  this->CmdSaveConfig->SetCommandName("SaveConfig");
+  this->CmdSaveConfig->SetBlocking(this->CommandBlocking);
+  this->CmdSaveConfig->SetDeviceID(this->CommandIGTLIODeviceID);
+
+}
+
+//----------------------------------------------------------------------------
+vtkSlicerPlusRemoteLogic::~vtkSlicerPlusRemoteLogic()
+{
+  this->CmdGetCaptureDeviceIDs->RemoveAllObservers();
```
Comment (lassoan) [inline PlusRemote/Logic/vtkSlicerPlusRemoteLogic.cxx]: Is this necessary? Doesn't deleting the VTK object deletes all the observers? If not, then should we remove all observers in the destructor?

File: PlusRemote/Logic/vtkSlicerPlusRemoteLogic.cxx [hunk 2]
```diff
@@ -0,0 +1,580 @@
...
+
+//----------------------------------------------------------------------------
+void vtkSlicerPlusRemoteLogic::CreateDefaultParameterSet()
+{
+  if (!this->GetMRMLScene())
+  {
+    return;
+  }
+
+  std::vector<vtkMRMLNode*> nodes = std::vector<vtkMRMLNode*>();
```
Comment (lassoan) [inline PlusRemote/Logic/vtkSlicerPlusRemoteLogic.cxx]: you can delete  = std::vector<vtkMRMLNode*>()

File: PlusRemote/Logic/vtkSlicerPlusRemoteLogic.h [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+  void StopLiveVolumeReconstruction(vtkMRMLIGTLConnectorNode* connectorNode, std::string volumeReconstructorDeviceID, std::string outputFilename, std::string outputDevice, vtkCallbackCommand* callback = NULL);
+  /// Sends a command to Plus to update a static transform that specified by the config file
+  void UpdateTransform(vtkMRMLIGTLConnectorNode* connectorNode, vtkMRMLLinearTransformNode* transformNode, vtkCallbackCommand* callback = NULL);
+  /// Sends a command to Plus to save the current config file, including modified parameters such as updated transforms
+  void SaveConfigFile(vtkMRMLIGTLConnectorNode* connectorNode, std::string configFileName, vtkCallbackCommand* callback = NULL);
+
+  /// Parses the command reply to get the filename of the reconstructed volume. (Assumes the filename is the section of the string after the last "/" character)
+  std::string ParseFilenameFromMessage(std::string filename);
+  /// Returns the current date and time as a string using the specified format
+  std::string GetTimestampString(std::string format = "%Y%m%d_%H%M%S");
```
Comment (lassoan) [inline PlusRemote/Logic/vtkSlicerPlusRemoteLogic.h]: can be static?

File: PlusRemote/Logic/vtkSlicerPlusRemoteLogic.cxx [hunk 4]
```diff
@@ -0,0 +1,580 @@
...
+  }
+  connectorNode->SendCommand(command);
+}
+
+//---------------------------------------------------------------------------
+std::string vtkSlicerPlusRemoteLogic::GetTimestampString(std::string format)
+{
+  // Get the string representing the current time.
+  // Format specified by input string.
+  // Default format is: YearMonthDay_HourMinuteSecond.
```
Comment (lassoan) [inline PlusRemote/Logic/vtkSlicerPlusRemoteLogic.cxx]: common convention is to write default argument values in the cxx files as comments:

std::string vtkSlicerPlusRemoteLogic::GetTimestampString(std::string format /*  = "%Y%m%d_%H%M%S" */ )

I would recommend to follow this convention and remove this comment line.

File: PlusRemote/Logic/vtkSlicerPlusRemoteLogic.cxx [hunk 5]
```diff
@@ -0,0 +1,580 @@
...
+  {
+    command->AddObserver(vtkSlicerOpenIGTLinkCommand::CommandCompletedEvent, callback);
+  }
+  connectorNode->SendCommand(command);
+}
+
+//---------------------------------------------------------------------------
+std::string vtkSlicerPlusRemoteLogic::GetTimestampString(std::string format)
+{
+  // Get the string representing the current time.
```
Comment (lassoan) [inline PlusRemote/Logic/vtkSlicerPlusRemoteLogic.cxx]: Move method documentation to header file

File: PlusRemote/Logic/vtkSlicerPlusRemoteLogic.cxx [hunk 6]
```diff
@@ -0,0 +1,580 @@
...
+  this->GetMRMLScene()->GetNodesByClass("vtkMRMLPlusRemoteNode", nodes);
+
+  if (nodes.empty())
+  {
+    this->GetMRMLScene()->AddNewNodeByClass("vtkMRMLPlusRemoteNode");
+  }
+}
+
+//---------------------------------------------------------------------------
+void vtkSlicerPlusRemoteLogic::GetCaptureDeviceIDs(vtkMRMLIGTLConnectorNode* connectorNode, vtkCallbackCommand* callback)
```
Comment (lassoan) [inline PlusRemote/Logic/vtkSlicerPlusRemoteLogic.cxx]: I don't think multiple modules can call GetCaptureDeviceIDs independently, as there is only a single CmdGetCaptureDeviceIDs command. So, we can simplify things and also make it more robust:
- remove vtkCallbackCommand callback argument and just update an associated PlusRemote MRML node with the result; the node will invoke Modified event (but you may also invoke some additional custom modified event that modules can observe)
- if a command is already in progress then return with an error

This is applicable to all commands.

File: PlusRemote/MRML/vtkMRMLPlusRemoteNode.cxx [hunk 7]
```diff
@@ -0,0 +1,616 @@
...
+      radius[i] = (dimensions[i] * spacing) / 2;
+    }
+    roiNode->SetRadiusXYZ(radius);
+  }
+  this->Modified();
+  this->EndModify(wasModifying);
+}
+
+//----------------------------------------------------------------------------
+void vtkMRMLPlusRemoteNode::WriteXML(ostream& of, int nIndent)
```
Comment (lassoan) [inline PlusRemote/MRML/vtkMRMLPlusRemoteNode.cxx]: Use new macros for read/write XML, print, and copy attributes:
vtkMRMLWriteXML…
vtkMRMLReadXML…
vtkMRMLCopy…
vtkMRMLPrint...

File: PlusRemote/Logic/vtkSlicerPlusRemoteLogic.cxx (hunk 6 above)
Comment (Sunderlandkyl) [reply PlusRemote/Logic/vtkSlicerPlusRemoteLogic.cxx]: That's a good idea. It also solves the problem of accessing the commands from Python.

File: PlusRemote/MRML/vtkMRMLPlusRemoteNode.cxx (hunk 7 above)
Comment (Sunderlandkyl) [reply PlusRemote/MRML/vtkMRMLPlusRemoteNode.cxx]: I used the macros where possible, but I couldn't adapt some attributes that are stored in std::vector.

**Comment by Sunderlandkyl** [comment]
This can be merged once https://github.com/Slicer/Slicer/pull/973 has been integrated.

**Comment by Sunderlandkyl** [comment]
OK, the required pull requests have been merged, and I've tested the build on both Mac and Windows.
This can be merged unless there are any issues/concerns/potential improvements.

**Pull request merged**